### ISSUE TITLE
`-fwrapv` incorrectly disables `-fsanitize=undefined` catching `-2147483648 % -1`

### ISSUE THREAD

**Issue by dzaima** [llvm/llvm-project#200152]
This program:

```c
int main() {
    int a = 1U << 31;
    int b = a % -1;
}
```

compiled with `-fwrapv -fsanitize=undefined`, crashes with `UndefinedBehaviorSanitizer: FPE on unknown address`, instead of properly detecting and reporting overflow as it does when ran without `-fwrapv` (note that `-fwrapv` does not prevent this operation being UB - the program still `SIGFPE`s when compiled with just `-fwrapv`).

And, combined with `-O1` or greater, will result in no sanitizer warnings at all.

The same applies to division:

```c
int main() {
    int a = 1U << 31;
    int b = a / -1;
}
```

https://godbolt.org/z/h9GK47fK7

**Comment by Sir-Walrus** [comment]
I'd expect ~~-2147483648 % -1 = -2147483648~~ -2147483648 / -1 = -2147483648 with -fwrapv, not a ubsan report.

But I do agree that SIGFPE not prefixed with a ubsan complaint is a bug.

e: fixed typoed operator

**Comment by dzaima** [comment]
> I'd expect -2147483648 % -1 = -2147483648 with -fwrapv, not a ubsan report.

Perhaps you mean `-2147483648 / -1 = -2147483648`, not modulo? `-2147483648 % -1`, if anything, should be `0` (as it is if you promote its args to 64-bit; and also what RISC-V and ARM, which have wrapping division, produce)

For what it's worth, gcc does do UB/SIGFPE for both the division and modulo tests with `-fwrapv`, and does report it as an error on `-fsanitize=undefined`.

**Comment by pinskia** [comment]
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=30484

**Comment by efriedma-quic** [comment]
See also #194474.
